### PR TITLE
Add simpleperf type to perf TraceCollector

### PR DESCRIPTION
**Pull request by robF8**

### CODE REVIEW

File: devlib/trace/perf.py [hunk 1]
```diff
@@ -16,34 +16,48 @@
...
+
 
 class PerfCollector(TraceCollector):
     """
     Perf is a Linux profiling with performance counters.
+    Simpleperf is an Android profiling tool woth performance counters.
+
+    It is highly recomended to use perf_type = simpleperf when using this instrument
+    on android devices, since it recognises android symbols in record mode and is much more stable
+    when reporting record .data files.
```
Comment (setrofim) [inline devlib/trace/perf.py]: Would be useful to provide a link to simpleperf documentation/wiki here.

File: devlib/trace/perf.py [hunk 2]
```diff
@@ -65,46 +80,61 @@ class PerfCollector(TraceCollector):
...
+                 target,
+                 perf_type='perf',
+                 command='stat',
                  events=None,
                  optionstring=None,
+                 report_options=None,
                  labels=None,
                  force_install=False):
         super(PerfCollector, self).__init__(target)
+        self.perf_type =perf_type
```
Comment (setrofim) [inline devlib/trace/perf.py]: - Fix spacing around `=`.
- Should validate that `pref_type` is one of the allowed values.

File: devlib/trace/perf.py [hunk 3]
```diff
@@ -65,46 +80,61 @@ class PerfCollector(TraceCollector):
...
             self.events = [self.events]
         if not self.labels:
             self.labels = ['perf_{}'.format(i) for i in range(len(self.optionstrings))]
         if len(self.labels) != len(self.optionstrings):
             raise ValueError('The number of labels must match the number of optstrings provided for perf.')
+        if command in ['stat', 'record']:
+            self.command = command
+        else:
+            raise ValueError('Unsupported perf command, must be stat or record')
+
```
Comment (setrofim) [inline devlib/trace/perf.py]: Remove the second blank line.

File: devlib/trace/perf.py [hunk 4]
```diff
@@ -65,46 +80,61 @@ class PerfCollector(TraceCollector):
...
         self.commands = self._build_commands()
 
     def reset(self):
-        self.target.killall('perf', as_root=self.target.is_rooted)
+        self.target.killall(self.perf_type, as_root=self.target.is_rooted)
         for label in self.labels:
-            filepath = self._get_target_outfile(label)
-            self.target.remove(filepath)
+            filepath = self._get_target_file(label, 'data')
+            #self.target.remove(filepath)
```
Comment (setrofim) [inline devlib/trace/perf.py]: Should be uncommented?

File: devlib/trace/perf.py [hunk 5]
```diff
@@ -65,46 +80,61 @@ class PerfCollector(TraceCollector):
...
+            filepath = self._get_target_file(label, 'rpt')
+            #self.target.remove(filepath)
```
Comment (setrofim) [inline devlib/trace/perf.py]: Should be uncommented?

File: devlib/trace/perf.py [hunk 6]
```diff
@@ -113,29 +143,67 @@ def stop(self):
     # pylint: disable=arguments-differ
     def get_trace(self, outdir):
         for label in self.labels:
-            target_file = self._get_target_outfile(label)
-            host_relpath = os.path.basename(target_file)
-            host_file = _f(os.path.join(outdir, host_relpath))
-            self.target.pull(target_file, host_file)
+            # Sometime perf/simpleperf takes a while to write to .data file after kill signal
```
Comment (setrofim) [inline devlib/trace/perf.py]: *Sometimes

File: devlib/trace/perf.py [hunk 7]
```diff
@@ -113,29 +143,67 @@ def stop(self):
...
+            time.sleep(5)
```
Comment (setrofim) [inline devlib/trace/perf.py]: Remove long fixed-length sleep -- wait on a condition (e.g. probe in a loop with a short sleep between iterations).

File: devlib/trace/perf.py [hunk 8]
```diff
@@ -113,29 +143,67 @@ def stop(self):
...
+            if self.command == 'record':
+                report_command = self._build_perf_report_command(self.report_options, label)
+                self.target.kick_off(report_command, as_root=True)
+                # For larger records, report file takes a while to generate
+                time.sleep(5)
```
Comment (setrofim) [inline devlib/trace/perf.py]: Remove long fixed-length sleep -- wait on a condition (e.g. probe in a loop with a short sleep between iterations).

File: devlib/trace/perf.py [hunk 9]
```diff
@@ -113,29 +143,67 @@ def stop(self):
...
 
+    def _get_target_file(self, label, extension):
+        return self.target.get_workpath('{}.{}'.format(label, extension))
+
     def _build_commands(self):
         commands = []
         for opts, label in zip(self.optionstrings, self.labels):
-            commands.append(self._build_perf_command(opts, self.events, label))
+            if self.command == 'stat':
+                commands.append(self._build_perf_command(opts, self.events, label))
```
Comment (setrofim) [inline devlib/trace/perf.py]: rename `_build_perf_command` to `_build_perf_stat_command`

File: devlib/trace/perf.py [hunk 10]
```diff
@@ -113,29 +143,67 @@ def stop(self):
...
+        return command
+
+    def _pull_target_file_to_host(self, label, extension, outdir):
+        target_file = self._get_target_file(label, extension)
+        host_relpath = os.path.basename(target_file)
+        host_file = _f(os.path.join(outdir, host_relpath))
+        self.target.pull(target_file, host_file)
+    
+    @staticmethod
+    def _get_default_events(perf_type):
```
Comment (setrofim) [inline devlib/trace/perf.py]: Just use a dict for this instead?

File: devlib/trace/perf.py [hunk 11]
```diff
@@ -16,34 +16,48 @@
...
+PERF_DEFAULT_EVENTS = [
     'migrations',
     'cs',
 ]
 
+SIMPLEPERF_DEFAULT_EVENTS = [
+    'raw-cpu-cycles',
+    'raw-l1-dcache',
+    'raw-l1-dcache-refill',
+    'raw-instructions-retired',
```
Comment (marcbonnici) [inline devlib/trace/perf.py]: `raw-instructions-retired` is an invalid event on my device (nexus5) 
Are there a set of predefined values that should be available on all devices?

Comment (marcbonnici) [inline devlib/trace/perf.py]: Do you know if there is a way to validate the events on the target? Currently the workload will execute and the only indication that the events are not valid is inside the simpleperf output file?

File: devlib/trace/perf.py (hunk 1 above)
Comment (robF8) [reply devlib/trace/perf.py]: Thank you, added link on both WA PR and here.

File: devlib/trace/perf.py (hunk 2 above)
Comment (robF8) [reply devlib/trace/perf.py]: Thanks, updated.

File: devlib/trace/perf.py (hunk 3 above)
Comment (robF8) [reply devlib/trace/perf.py]: Removed

File: devlib/trace/perf.py (hunk 4 above)
Comment (robF8) [reply devlib/trace/perf.py]: Yes it should, that was careless sorry

File: devlib/trace/perf.py (hunk 5 above)
Comment (robF8) [reply devlib/trace/perf.py]: Fixed also

File: devlib/trace/perf.py (hunk 7 above)
Comment (robF8) [reply devlib/trace/perf.py]: Let me know what you think about my solution to this.

File: devlib/trace/perf.py (hunk 9 above)
Comment (robF8) [reply devlib/trace/perf.py]: Fixed

File: devlib/trace/perf.py (hunk 10 above)
Comment (robF8) [reply devlib/trace/perf.py]: Updated to use dict

File: devlib/trace/perf.py [hunk 12]
```diff
@@ -16,34 +16,48 @@
...
+]
+
 
 class PerfCollector(TraceCollector):
     """
     Perf is a Linux profiling with performance counters.
+    Simpleperf is an Android profiling tool woth performance counters.
```
[the remark this answers is not in the document]
Comment (robF8) [reply devlib/trace/perf.py]: Fixed

File: devlib/trace/perf.py [hunk 13]
```diff
@@ -65,46 +80,61 @@ class PerfCollector(TraceCollector):
...
+                 perf_type='perf',
+                 command='stat',
                  events=None,
                  optionstring=None,
+                 report_options=None,
                  labels=None,
                  force_install=False):
         super(PerfCollector, self).__init__(target)
+        self.perf_type =perf_type
         self.events = events if events else DEFAULT_EVENTS
```
[the remark this answers is not in the document]
Comment (robF8) [reply devlib/trace/perf.py]: Updated to use dict

File: devlib/trace/perf.py (hunk 11 above)
Comment (robF8) [reply devlib/trace/perf.py]: Yes that was a last minute typo sorry, should have been instruction instead of instructions. I have updated

Comment (robF8) [reply devlib/trace/perf.py]: That's a good point. Is it worth just calling perf list and then checking them against the events to make sure user hasn't misspelled an event?

**Comment by robF8** [comment]
Added updates based on feedback

File: devlib/trace/perf.py [hunk 14]
```diff
@@ -113,29 +151,87 @@ def stop(self):
...
+                current_tries += 1
+            else:
+                if current_tries >= max_tries:
+                    self.logger.warning('''writing {}.data file took longer than expected, 
+                                        file may not have written correctly'''.format(label))
+                data_file_finished_writing = True
+        report_command = self._build_perf_report_command(self.report_options, label)
+        self.target.execute(report_command)
+
+    def _validate_events(self, events):
```
Comment (marcbonnici) [inline devlib/trace/perf.py]: Just checked this with an original perf agenda and the old default events `migrations` and `cs` are not in the available events and do not pass the validation however `cpu-migrations` and `context-switches` are present. Do you know if perf supports additional aliases that are not present in the list?

Comment (robF8) [reply devlib/trace/perf.py]: I don't know off the top of my head but will investigate now and find out.

**Comment by robF8** [comment]
Updated to accept aliases for available events

**Pull request merged**